### DISCUSSION TITLE
単体テストはどこまで書いてますか？

### DISCUSSION THREAD

**Discussion by komura-c** [StartupAngular/discussions#36] [General]

**Comment by rettar5** [comment]
セッション1

**Reply by rettar5** [reply]
## ざっくり
- あんまりかけてないが、ロジックはサーバー側(API)によっていて、そっちで担保してます
- ↑全くおんなじ感じ
    - フロント側には書いてないが、テスト採用する動きをしている、今書き始めてる感じ
    - サーバ側で担保してる
- Angularの場合、jasmine使ってHTMLのUIの部分とサービスの部分とコンポーネントストアの3部分でテスト書いている
    - 組織的にテスト書いていこうというところまではいけてない
    - 人によって書いてたり書いてなかったり
    - 4年ぐらい関わってるプロダクトでは、最初はテストはかけてなかったが事故を経験してテストの必要性を感じた
    - どういうテストの書き方がいいのか研究して、少しずつ充実させていっている
- 前職のはなし
    - コンポーネントのところはコストが高い
    - ロジック部分(サービスクラス)に対して単体テストを書くようにしてる
- もともとe2eででかいのを書いてた
    - なくなる & 壊れやすい
    - unit testはまだ軌道に乗れてない
- 最近になって新しく作るコンポーネントについては書いていく方針になった
    - methodやサービスにはテスト書いているが、UIについてはひとによりけり
    - 前職ではUI部分もテストしていた

## HTMLのUIのテストをどこまでテストしたらいいのかわからない
- 最大文字数のテストとか
    - reactive formだとvalidatorを自分で作れるので、validator functionsまではテストしてた
    - それが表示されてるところまではテストしてない
    - ついでに書けるレベルだったのでチャットアプリではテストしてた
    - e2eは成功系だけ見てた
        - 単体テストだと境界や異常系まで見れるけど
    - AngularだとUIとコンポーネントが密結合なので、ロジックをどこに出すか次第なきがする
        - コンポーネントでテストしづらいから別で切り出す、っていうのがどのくらいの粒度によるか次第？

## コンポーネントどれくらいの重さ？
- あんまり意識できてない
    - 切り出しとかはやりつつ、チームによっては切り離せずそのまま書いてるところとかもある
        - コンポーネントに使うものはコンポーネントに書いてたり
    - 人が増えるとリファクタできない状況が続いたり
- pageとかの単位でコンポーネント切ったりする？
    - 画面によるがページで切ってるところもある
    - サービス作って切り出してるところもある
- コンポーネントストア
    - コンポーネントに紐づくストアにロジックを書いていたりする
    - https://kasaharu.hatenablog.com/entry/20210414/1618399842
    - componentに紐づくサービスクラスを作るイメージ
        - 1つのコンポーネントに対して1サービス
        - ディレクトリは？
            - コンポーネントと横並びでストアがある
        - サービス作ろうと思うが、ディレクトリ構成悩んで気が引ける
        - ページの配下にsharedを作ってるのが面倒でサービスを作らないことある
            - sharedに置くのはページまたいで使うサービス
    - ロジック部分をmock化できるので、テストがしやすくなる
    - いかに責務を分離するかがテストのしやすさにつながりそう
        - UIの責務はコンポーネントの中でロジックに渡すまで
        - そこから先はコンポーネントストア、サービスでテストする
        - フォーカスする部分を分離できる
            - テストケースを簡単にできる

## 単体テスト書くときにmodule import大変じゃない？
- 必要なモジュール読み込むのどうしてる？
    - ごそっと入れてる感じ？
    - UI変わるたびにメンテするのが大変で困ってる
- 初めてjasmine書いたときに知らなくて苦労した
    - 普通のコンポーネントと同じだが…
    - テストが本当に必要なコンポーネントだけを定義していくと、最小・最速でテストできるはず
        - 昔はテストするごとにimportするべきだったが、今はテストが高速化してる
        - importした先がさらに依存してると管理が大変
        - UIやcommonでmoduleを分割して、UIのときはUIのmoduleだけを入れればOKって感じにしてる

## テスト用のコンポーネント・ディレクティブって作ってる？
- e2eだとテストしやすくするディレクティブがあった
- 単体テストにもそういう仕組あったりする？
    - HTMLのname属性指定でテストしてる

**Reply by rettar5** [reply]
https://ngrx.io/guide/component-store

**Comment by scrpgil** [comment]
セッション2

**Reply by scrpgil** [reply]
# 概要
- テストはプルリクを作った時にセットに書くようになっている。コンポーネントなどを作ったときには、必ず作るのが前提。ただ、完璧に書くと大変なので、in-out周りのテストがメイン。それ以上は関わる。
- フロントエンドの単体テストはかけていない。やりたいという話は出つつ、書けていない。e2eをやっちゃえば良いんじゃない？という話になっている。cypressというe2eのフレームワークを使おうかと考えている。
- 最近テストを書き始めた。こりすぎると工数もとられてしまう。
- 新機能の開発の場合は、絶対テストは入れる。既存の改修のような時間がかかりそうな場合は、QA頼み。ただ、テストの粒度はかなり細かくやっていた。
- 書いてない勢力。新しくPJ立ち上げると、jestとかjasmine入ってる。書かないと不安になる。e2eを検討したけど入れなかった。

## 深堀り
- どういうタイミングがきたから単体テスト入れようという話になったのか？
  -  複数プロダクト、複数チームにわけようというタイミングで、既存のコード資産の流用性を高めようという話が発端で単体テスト入れようという風潮が一気に高まった。
  - エンジニアチームとして、テストを入れようという流れになった。ちょこちょこ不具合とかに気づかないことがあり、テスト入れようという流れになった。lintもそのタイミングで入れた。
  - 若いプロダクトで、機能をガンガン入れてく段階だったので、テストなしでやっていた。
  - 7~8年ぐらい前に作ったプロダクトを育て上げている。バックエンドのエンジニアって、テスト書かなければという風潮が多かった。その流れで、フロントエンドもテスト書くのが当たり前だった。
     - bugsnag、ランタイムエラーをとってくれるツール。
  - 前いた現場は、新機能を作るタイミングで必ず入れるのが普通。それで、落ちるとエンジニア側の気付きにもなった。仕様書としてのテストという位置づけも多くあった。
- テストフレームワーク、jest、karma、jasmine、selenium、cypress。
  - karma、jasmineを使っている。
  - testing-library：https://testing-library.com/docs/angular-testing-library/examples/
  - ng-mocks
  - jest + Angular Testing Library + ng-mocksを使っている。
  - cypress。
- e2eまでやりますか？

**Reply by komura-c** [reply]
https://www.bugsnag.com/

**Reply by adachi-tsukasa** [reply]
https://sinonjs.org/

**Reply by komura-c** [reply]
ng-japanテスト会
https://www.youtube.com/watch?v=ZTToIQF7zyc
